### PR TITLE
feat(lib): add account types for different signing providers

### PR DESCRIPTION
**Pull request by lucasfernog**
# Description of change

Introduces account type so different signing providers are supported (e.g. stronghold, ledger, plain mnemonic...) 

## Links to any relevant issues

#80 

## Type of change

- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

Unit tests.

## Change checklist

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have checked that new and existing unit tests pass locally with my changes

### CODE REVIEW

File: src/signing/stronghold.rs [hunk 1]
```diff
@@ -0,0 +1,69 @@
+use crate::account::{account_id_to_stronghold_record_id, Account};
```
Comment (thibault-martinez) [inline src/signing/stronghold.rs:12]: Missing copyright :p

File: src/signing/mod.rs [hunk 2]
```diff
@@ -0,0 +1,94 @@
+use std::{
```
Comment (thibault-martinez) [inline src/signing/mod.rs:12]: Missing copyright :p

File: src/account/mod.rs [hunk 3]
```diff
@@ -79,9 +81,19 @@ impl<'a> AccountInitialiser<'a> {
...
             storage_path,
+            #[cfg(feature = "stronghold")]
+            account_type: Some(SignerType::Stronghold),
+            #[cfg(not(feature = "stronghold"))]
+            account_type: None,
         }
     }
 
+    /// Sets the account type.
+    pub fn account_type(mut self, account_type: SignerType) -> Self {
```
Comment (thibault-martinez) [inline src/account/mod.rs]: Seems weird to me that account_type is of type SignerType

File: src/signing/mod.rs [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+use once_cell::sync::OnceCell;
+use serde::{Deserialize, Serialize};
+use slip10::BIP32Path;
+
+mod stronghold;
+use self::stronghold::StrongholdSigner;
+
+type BoxedSigner = Box<dyn Signer + Sync + Send>;
+type Signers = Arc<RwLock<HashMap<SignerType, BoxedSigner>>>;
+static SIGNER_INSTANCES: OnceCell<Signers> = OnceCell::new();
```
Comment (thibault-martinez) [inline src/signing/mod.rs]: Nit but I think `SIGNERS_INSTANCE` is more appropriate as there is one instance of signers and not instances of signer.

File: src/signing/mod.rs [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    /// Initialises an account.
+    fn init_account(&self, account: &Account, mnemonic: Option<String>) -> crate::Result<String>;
+    /// Generates an address.
+    fn generate_address(
+        &self,
+        account: &Account,
+        index: usize,
+        internal: bool,
+    ) -> crate::Result<iota::Address>;
+    /// Sign message.
```
Comment (thibault-martinez) [inline src/signing/mod.rs]: For consistency
```suggestion
    /// Signs message.
```

File: src/signing/mod.rs [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+/// Sets the signer interface for the given type.
+pub fn set_signer<S: Signer + Sync + Send + 'static>(signer_type: SignerType, signer: S) {
+    let mut instances = SIGNER_INSTANCES
+        .get_or_init(default_signers)
+        .write()
+        .unwrap();
+    instances.insert(signer_type, Box::new(signer));
+}
+
+/// gets the signer interface.
```
Comment (thibault-martinez) [inline src/signing/mod.rs]:
```suggestion
/// Gets the signer interface.
```

File: src/signing/stronghold.rs [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+  fn init_account(&self, account: &Account, mnemonic: Option<String>) -> crate::Result<String> {
+    let stronghold_account_res: crate::Result<stronghold::Account> =
+      crate::with_stronghold_from_path(account.storage_path(), |stronghold| {
+        let created_at_timestamp: u128 = account.created_at().timestamp().try_into().unwrap(); // safe to unwrap since it's > 0;
+        let account = match mnemonic {
+          Some(mnemonic) => stronghold.account_import(
+            Some(created_at_timestamp),
+            Some(created_at_timestamp),
+            mnemonic,
+            Some("password"),
```
Comment (thibault-martinez) [inline src/signing/stronghold.rs]: I lack overall knowledge of wallet.rs so: what is this ?

**Review by thibault-martinez — COMMENTED** [review]
LGTM! Some nits here and there :)

File: src/signing/stronghold.rs (hunk 7 above)
Comment (lucasfernog) [reply src/signing/stronghold.rs]: do not worry about this, it's a dummy mess we had with the old stronghold :P

File: src/account/mod.rs (hunk 3 above)
Comment (lucasfernog) [reply src/account/mod.rs]: changed it to be a little more consistent.

**Review by cvarley100 — APPROVED** [review]

**Pull request merged**